README.md
# Always Allow Admin Comments

A plugin for WordPress that allows an admin user (when logged in) to always be able to comment on a post, even if comments are closed for the post.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/always-allow-admin-comments/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress


## Additional Documentation

* See [readme.txt](https://github.com/coffee2code/always-allow-admin-comments/blob/master/readme.txt) for additional usage information.
* See [DEVELOPER-DOCS.md](DEVELOPER-DOCS.md) for developer-related documentation on provided hooks.
* See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/always-allow-admin-comments/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/always-allow-admin-comments/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/always-allow-admin-comments/reviews/) for it in the WordPress Plugin Directory or considerating a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 1.3.2 _(2021-10-14)_
* New: Add DEVELOPER-DOCS.md and move hooks documentation into it
* Change: Note compatibility through WP 5.8+
* Change: Tweak installation instruction
* Unit tests:
    * Change: Explicitly test a custom post type that supports comments and one that does not
    * Change: Restructure unit test directories
        * Change: Move `bin` into `tests/`
        * Change: Move `tests/bootstrap.php` into `tests/phpunit/`
        * Change: Move `tests/test-*.php` into `tests/phpunit/tests/`
    * Change: Remove 'test-' prefix from unit test file
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
    * Change: In bootstrap, store path to plugin file constant
    * Change: In bootstrap, add backcompat for PHPUnit pre-v6.0

## 1.3.1 _(2021-04-04)_
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)

## 1.3 _(2020-05-12)_
* Change: Add customization for post type support
    * New: Add filter `c2c_always_allow_admin_comments_post_types` to allow customizing supported post types
    * New: Add `get_post_types()` to get the list of supported post types
    * Change: Enhance `can_show_ui()` to check if post's post type is supported
* Change: Change label for setting
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add more items to the list)
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Unit tests:
    * New: Add tests for registering of hooks
    * Change: Remove unnecessary unregistering of hooks

## 1.2.2 _(2019-12-28)_
* Unit tests:
    * New: Add test to verify plugin hooks `plugins_loaded` action to initialize itself
    * New: Add unit tests for functions `can_show_ui()`, `comments_open_for_admin()`
* Change: Note compatibility through WP 5.3+
* Change: Add an FAQ to clarify that the checkbox to disable admin comments is only available for the classic editor
* Change: Update copyright date (2020)

## 1.2.1 _(2019-06-18)_
* Change: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Note compatibility through WP 5.2+
* Change: Add link to CHANGELOG.md in README.md
* Fix: Correct typo in GitHub URL
* Fix: Use full path to CHANGELOG.md in the Changelog section of readme.txt

## 1.2 _(2019-03-20)_

### Highlights:

This release mainly adds support for the block editor (aka Gutenberg) via an interim metabox "Prevent Admin Comments". (The block editor does not currently support adding the plugin's checkbox in the "Discussion" panel as expected.) Otherwise, there were a number of behind-the-scenes code and documentation changes.

### Details:

* New: Add `do_meta_box()` for interim metabox for block editor (aka Gutenberg) support
* New: Add `can_show_ui()` to determine if the UI should generally be shown
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hook
* New: Add .gitignore file
* New: Add screenshot of new block editor metabox
* Change: Modify how the meta field is registered:
    * Use `register_post_meta()` instead of `register_meta()`, if available
    * Explicitly register meta for each post type that supports having comments
    * Set `show_in_rest` to true
    * Set `type` to boolean
    * Add callbacks to `auth_callback` and `sanitize_callback`
    * Register meta on `init` instead of `plugins_loaded`
* Change: Revamp `display_option()` to operate in multiple contexts
    * Make `$post` argument optional, defaulting to global post
    * Improve styling when metabox appears on the right
    * Use `sprintf()` instead of string concatenation to build output strings
    * Move styles into `style` tag instead of being inline styles
* Change: Add README link to plugin's page in Plugin Directory
* Change: Split paragraph in README.md's "Support" section into two
* Change: Initialize plugin on 'plugins_loaded' action instead of on load
* Change: Merge `plugins_loaded()` into constructor
* Change: Tweak plugin description
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Note compatibility through WP 5.1+
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS

## 1.1.1 _(2017-11-07)_
* New: Add README.md
* Change: Minor tweak to plugin description
* Change: Minor whitespace changes to unit test bootstrap
* Change: Add GitHub link to readme
* Change: Note compatibility through WP 4.9+
* Change: Update copyright date (2018)

## 1.1 _(2017-01-23)_
* Change: Register meta field via `register_meta()`.
    * Add own `register_meta()`
    * Remove `hide_meta()` in favor of use of `register_meta()`
* Change: Sanitize meta key name when used as input attribute (it's not a user input value so no security issue existed).
* Change: Enable more error output for unit tests.
* Change: Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable.
* Change: Minor readme.txt documentation tweaks.
* Change: Note compatibility through WP 4.7+.
* Change: Remove support for WordPress older than 4.6 (should still work for earlier versions)
* Change: Minor readme improvements.
* Change: Update copyright date (2017).

## 1.0 _(2016-03-08)_
* Initial public release


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Always_Allow_Admin_Comments
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/always-allow-admin-comments.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Always_Allow_Admin_Comments
 */

define( 'ALWAYS_ALLOW_ADMIN_COMMENTS_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/always-allow-admin-comments.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require ALWAYS_ALLOW_ADMIN_COMMENTS_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/test-always-allow-admin-comments.php
<?php

defined( 'ABSPATH' ) or die();

class test_AlwaysAllowAdminComments extends WP_UnitTestCase {

	private static $meta_key = 'c2c_always_allow_admin_comments';

	public function setUp() {
		parent::setUp();
		// This shouldn't be necessary.
		do_action( 'init' );
	}

	public function tearDown() {
		parent::tearDown();
		$this->unset_current_user();
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	private function create_user( $role, $set_as_current = true ) {
		$user_id = $this->factory->user->create( array( 'role' => $role ) );
		if ( $set_as_current ) {
			wp_set_current_user( $user_id );
		}
		return $user_id;
	}

	// helper function, unsets current user globally. Taken from post.php test.
	private function unset_current_user() {
		global $current_user, $user_ID;

		$current_user = $user_ID = null;
	}


	//
	//
	// FUNCTIONS FOR HOOKING ACTIONS/FILTERS
	//
	//


	public function disable_admin_commenting_on_specified_post( $status, $post_id ) {
		$post = get_post( $post_id );
		return 'Admin cannot comment' === $post->post_title ? true : $status;
	}

	public function enable_admin_commenting_on_specified_post( $status, $post_id ) {
		$post = get_post( $post_id );
		return 'Admin cannot comment' === $post->post_title ? false : $status;
	}


	//
	//
	// DATA PROVIDERS
	//
	//


	public static function get_default_hooks() {
		// Function names prefixed with "::" are treated as class methods
		// instead of object methods.
		return array(
			array( 'filter', 'comments_open',                        'comments_open_for_admin', 20 ),
			array( 'action', 'post_comment_status_meta_box-options', 'display_option',          10 ),
			array( 'action', 'save_post',                            'save_setting',            10 ),
			array( 'action', 'do_meta_boxes',                        'do_meta_box',             10 ),
			array( 'action', 'init',                                 '::register_meta',         10 ),
		);
	}


	//
	//
	// TESTS
	//
	//


	public function test_class_exists() {
		$this->assertTrue( class_exists( 'c2c_AlwaysAllowAdminComments' ) );
	}

	public function test_version() {
		$this->assertEquals( '1.3', c2c_AlwaysAllowAdminComments::get_instance()->version() );
	}

	public function test_instance_object_is_returned() {
		$this->assertTrue( is_a( c2c_AlwaysAllowAdminComments::get_instance(), 'c2c_AlwaysAllowAdminComments' ) );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_filter( 'plugins_loaded', array( 'c2c_AlwaysAllowAdminComments', 'get_instance' ) ) );
	}

	/**
	 * @dataProvider get_default_hooks
	 */
	public function test_default_hooks( $hook_type, $hook, $function, $priority ) {
		$callback = 0 === strpos( $function, '::' )
			? array( 'c2c_AlwaysAllowAdminComments', substr( $function, 2 ) )
			: array( c2c_AlwaysAllowAdminComments::get_instance(), $function );

		$prio = $hook_type === 'action' ?
			has_action( $hook, $callback ) :
			has_filter( $hook, $callback );
		$this->assertNotFalse( $prio );
		if ( $priority ) {
			$this->assertEquals( $priority, $prio );
		}
	}

	/*
	 * get_post_types()
	 */

	public function test_get_post_types() {
		$this->assertEquals( array( 'post', 'page', 'attachment' ), c2c_AlwaysAllowAdminComments::get_post_types() );
	}

	/*
	 * filter: c2c_always_allow_admin_comments_post_types
	 */

	 public function test_filter_c2c_always_allow_admin_comments_post_types() {
		add_filter( 'c2c_always_allow_admin_comments_post_types', function( $post_types ) {
			if ( false !== $i = array_search( 'page', $post_types ) ) {
				unset( $post_types[ $i ] );
			}
			return $post_types;
		} );

		$this->assertEquals( array( 'post', 'attachment' ), c2c_AlwaysAllowAdminComments::get_post_types() );
	}

	/*
	 * Core functionality via comments_open()
	 */

	 public function test_admin_can_comment_when_comments_closed() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'closed' ) );
		$user_id = $this->create_user( 'administrator' );

		$this->assertTrue( comments_open( $post_id ) );
	}

	public function test_non_admin_cannot_comment_when_comments_closed() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'closed' ) );

		$user_id = $this->create_user( 'editor' );
		$this->assertFalse( comments_open( $post_id ) );

		$user_id = $this->create_user( 'subscriber' );
		$this->assertFalse( comments_open( $post_id ) );

		$this->unset_current_user();
		$this->assertFalse( comments_open( $post_id ) );
	}

	public function test_anyone_can_comment_when_comments_open() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open' ) );

		$user_id = $this->create_user( 'administrator' );
		$this->assertTrue( comments_open( $post_id ) );

		$user_id = $this->create_user( 'editor' );
		$this->assertTrue( comments_open( $post_id ) );

		$user_id = $this->create_user( 'subscriber' );
		$this->assertTrue( comments_open( $post_id ) );

		$this->unset_current_user();
		$this->assertTrue( comments_open( $post_id ) );
	}

	public function test_admin_can_comment_on_old_posts() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open', 'post_date' => '2010-10-01 13:12:25', 'post_date_gmt' => '2010-10-01 13:12:25' ) );

		update_option( 'close_comments_for_old_posts', '1' );
		$x = get_option( 'close_comments_for_old_posts' );
		$y = (int) get_option( 'close_comments_days_old' );
		// In case WP closed comments on older posts
		$this->assertEquals( true, $x );
		$this->assertFalse( comments_open( $post_id ) );

		update_option( 'close_comments_for_old_posts', '0' );
	}

	public function test_post_setting_can_disable_admin_commenting() {
		$post_id1 = $this->factory->post->create( array( 'post_title' => 'Admin cannot comment', 'comment_status' => 'closed' ) );
		$post_id2 = $this->factory->post->create( array( 'post_title' => 'Aaaa', 'comment_status' => 'closed' ) );
		$user_id = $this->create_user( 'administrator' );

		$obj = c2c_AlwaysAllowAdminComments::get_instance();
		$obj->set_admin_can_comment_on_post( false, $post_id1 );

		$this->assertFalse( comments_open( $post_id1 ) );
		$this->assertTrue(  comments_open( $post_id2 ) );
		$this->assertTrue(  $obj->is_admin_commenting_disabled( $post_id1 ) );
		$this->assertTrue(  $obj->is_admin_commenting_disabled( $post_id1, false ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id2 ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id2, false ) );
	}

	/*
	 * filter: c2c_always_allow_admin_comments_disable
	 */

	public function test_filter_can_disable_admin_commenting() {
		$post_id1 = $this->factory->post->create( array( 'post_title' => 'Admin cannot comment', 'comment_status' => 'closed' ) );
		$post_id2 = $this->factory->post->create( array( 'post_title' => 'Aaaa', 'comment_status' => 'closed' ) );
		$user_id = $this->create_user( 'administrator' );

		$this->assertTrue( comments_open( $post_id1 ) );
		$this->assertTrue( comments_open( $post_id2 ) );

		$obj = c2c_AlwaysAllowAdminComments::get_instance();
		add_filter( 'c2c_always_allow_admin_comments_disable', array( $this, 'disable_admin_commenting_on_specified_post' ), 10, 2 );

		$this->assertFalse( comments_open( $post_id1 ) );
		$this->assertTrue(  comments_open( $post_id2 ) );
		$this->assertTrue(  $obj->is_admin_commenting_disabled( $post_id1 ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id1, false ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id2 ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id2, false ) );
	}

	public function test_filter_overrides_post_setting() {
		$post_id1 = $this->factory->post->create( array( 'post_title' => 'Admin cannot comment', 'comment_status' => 'closed' ) );
		$post_id2 = $this->factory->post->create( array( 'post_title' => 'Aaaa', 'comment_status' => 'closed' ) );
		$user_id = $this->create_user( 'administrator' );

		$this->assertTrue( comments_open( $post_id1 ) );
		$this->assertTrue( comments_open( $post_id2 ) );

		// Directly disable admin commenting in the post, but enable it via filter.
		$obj = c2c_AlwaysAllowAdminComments::get_instance();
		$obj->set_admin_can_comment_on_post( false, $post_id1 );
		add_filter( 'c2c_always_allow_admin_comments_disable', array( $this, 'enable_admin_commenting_on_specified_post' ), 10, 2 );

		$this->assertTrue(  comments_open( $post_id1 ) );
		$this->assertTrue(  comments_open( $post_id2 ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id1 ) );
		$this->assertTrue(  $obj->is_admin_commenting_disabled( $post_id1, false ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id2 ) );
		$this->assertFalse( $obj->is_admin_commenting_disabled( $post_id2, false ) );
	}

	public function test_set_admin_can_comment_on_post() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'closed' ) );

		$this->assertFalse( comments_open( $post_id ) );

		$user_id = $this->create_user( 'administrator' );

		$this->assertTrue( comments_open( $post_id ) );

		// Directly disable admin commenting in the post, but enable it via filter.
		$obj = c2c_AlwaysAllowAdminComments::get_instance();
		$obj->set_admin_can_comment_on_post( true, $post_id );

		$this->assertTrue( comments_open( $post_id ) );

		$obj->set_admin_can_comment_on_post( false, $post_id );

		$this->assertFalse( comments_open( $post_id ) );
	}

	public function test_meta_is_registered() {
		$this->assertTrue( registered_meta_key_exists( 'post', self::$meta_key, 'post' ) );
	}

	/*
	 * can_show_ui()
	 */

	public function test_can_show_ui_for_admin() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open' ) );
		$this->create_user( 'administrator' );

		$this->assertTrue( c2c_AlwaysAllowAdminComments::get_instance()->can_show_ui( $post_id ) );
	}

	public function test_can_show_ui_for_non_admin() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open' ) );
		$this->assertFalse( c2c_AlwaysAllowAdminComments::get_instance()->can_show_ui( $post_id ) );

		$this->create_user( 'editor' );

		$this->assertFalse( c2c_AlwaysAllowAdminComments::get_instance()->can_show_ui( $post_id ) );
	}

	public function test_can_show_ui_for_unsupported_post_type() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open', 'post_type' => 'bogus' ) );
		$this->create_user( 'administrator' );

		$this->assertFalse( c2c_AlwaysAllowAdminComments::get_instance()->can_show_ui( $post_id ) );
	}

	/*
	 * comments_open_for_admin()
	 */

	public function test_comments_open_for_admin_with_open_comments_and_admin() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open' ) );
		$this->create_user( 'administrator' );

		$this->assertTrue( c2c_AlwaysAllowAdminComments::get_instance()->comments_open_for_admin( true, $post_id ) );
	}

	public function test_comments_open_for_admin_with_closed_comments_and_admin() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'closed' ) );
		$this->create_user( 'administrator' );

		$this->assertTrue( c2c_AlwaysAllowAdminComments::get_instance()->comments_open_for_admin( false, $post_id ) );
	}

	public function test_comments_open_for_admin_with_closed_comments_and_admin_comments_disabled() {
		$post_id = $this->factory->post->create( array( 'post_title' => 'Admin cannot comment', 'comment_status' => 'closed' ) );
		$this->create_user( 'administrator' );
		add_filter( 'c2c_always_allow_admin_comments_disable', array( $this, 'disable_admin_commenting_on_specified_post' ), 10, 2 );

		$this->assertFalse( c2c_AlwaysAllowAdminComments::get_instance()->comments_open_for_admin( false, $post_id ) );
	}

	public function test_comments_open_for_admin_with_open_comments_and_non_admin() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'open' ) );
		$this->create_user( 'editor' );

		$this->assertTrue( c2c_AlwaysAllowAdminComments::get_instance()->comments_open_for_admin( true, $post_id ) );
	}

	public function test_comments_open_for_admin_with_closed_comments_and_non_admin() {
		$post_id = $this->factory->post->create( array( 'comment_status' => 'closed' ) );
		$this->create_user( 'editor' );

		$this->assertFalse( c2c_AlwaysAllowAdminComments::get_instance()->comments_open_for_admin( false, $post_id ) );
	}

}


always-allow-admin-comments.php
<?php
/**
 * Plugin Name: Always Allow Admin Comments
 * Version:     1.3.2
 * Plugin URI:  https://coffee2code.com/wp-plugins/admin-can-always-comment/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: always-allow-admin-comments
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Allow an admin user (when logged in) to always be able to comment on a post, even if comments are closed for the post.
 *
 * Compatible with WordPress 4.6+ through 5.8+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/always-allow-admin-comments/
 *
 * @package Always_Allow_Admin_Comments
 * @author  Scott Reilly
 * @version 1.3.2
 */

/*
	Copyright (c) 2013-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_AlwaysAllowAdminComments' ) ) :

class c2c_AlwaysAllowAdminComments {

	/**
	 * The singleton instance of this class.
	 *
	 * @var c2c_AlwaysAllowAdminComments
	 */
	private static $instance;

	/**
	 * The meta key for storing an override to prevent admins from
	 * commenting on a particular post.
	 *
	 * @var string
	 */
	private static $setting_name = 'c2c_always_allow_admin_comments';

	/**
	 * Gets singleton instance, creating it if necessary.
	 *
	 * @since 1.0
	 */
	public static function get_instance() {
		if ( ! self::$instance ) {
			self::$instance = new self;
		}

		return self::$instance;
	}

	/**
	 * Returns version of the plugin.
	 *
	 * @since 1.0
	 */
	public static function version() {
		return '1.3.2';
	}

	/**
	 * The constructor.
	 *
	 * @since 1.0
	 */
	private function __construct() {
		// Load textdomain
		load_plugin_textdomain( 'always-allow-admin-comments' );

		add_filter( 'comments_open',                        array( $this, 'comments_open_for_admin' ), 20, 2 );
		add_action( 'post_comment_status_meta_box-options', array( $this, 'display_option' ) );
		add_action( 'save_post',                            array( $this, 'save_setting' ) );

		add_action( 'do_meta_boxes',                        array( $this, 'do_meta_box' ), 10, 3 );

		add_action( 'init',                                 array( __CLASS__, 'register_meta' ) );
	}

	/**
	 * Returns the support post types.
	 *
	 * @psince 1.3
	 *
	 * @return array
	 */
	public static function get_post_types() {
		/**
		 * Filters whether admin commenting has been explicitly disabled for the
		 * given post.
		 *
		 * @since 1.3
		 *
		 * @param array $post_types Supported post types. Default is all the post
		 *                          types that support 'comments' as a feature.
		 */
		return array_values( (array) apply_filters( 'c2c_always_allow_admin_comments_post_types', get_post_types_by_support( 'comments' ) ) );
	}

	/**
	 * Registers the post meta field.
	 *
	 * @since 1.1
	 */
	public static function register_meta() {
		$config = array(
			'type'              => 'boolean',
			'description'       => __( 'Prevent administrators from commenting?', 'always-allow-admin-comments' ),
			'single'            => true,
			'sanitize_callback' => function ( $value ) {
				return (bool) $value;
			},
			'auth_callback'     => function() {
				return current_user_can( 'edit_posts' );
			},
			'show_in_rest'      => true,
		);

		$post_types = self::get_post_types();
		if ( ! $post_types ) {
			return;
		}

		foreach ( $post_types as $post_type ) {
			if ( function_exists( 'register_post_meta' ) ) {
				register_post_meta( $post_type, self::$setting_name, $config );
			}
			// Pre WP 4.9.8 support
			else {
				$config['object_subtype'] = $post_type;
				register_meta( 'post', self::$setting_name, $config );
			}
		}
	}

	/**
	 * Register meta box.
	 *
	 * Due to the block editor's current lack of panel customization, it is not
	 * presently feasible to insert the checkbox into the Duscussion panel where
	 * it appropriately belongs. So until it is possible to do so, just put the
	 * field into a meta box. (The meta box is not needed in the classic editor.)
	 *
	 * @since 1.2
	 *
	 * @param string  $post_type The post type.
	 * @param string  $type      The mode for the meta box (normal, advanced, or side).
	 * @param WP_Post $post      The post.
	 */
	public function do_meta_box( $post_type, $type, $post ) {
		// Can the UI be shown?
		if ( ! $this->can_show_ui( $post ) ) {
			return;
		}

		$current_screen = get_current_screen();

		// Don't show unless in the block editor.
		if ( ! method_exists( $current_screen, 'is_block_editor' ) || ! $current_screen->is_block_editor() ) {
			return;
		}

		add_meta_box(
			'always-allow-admin-comments-div',
			__( 'Prevent Admin Comments', 'always-allow-admin-comments' ),
			array( $this, 'display_option' ),
			$post_type,
			'side',
			'core'
		);
	}

	/**
	 * Determines if the UI for the plugin can be shown.
	 *
	 * Criteria for the UI to be shown:
	 * - Current user is an administrator
	 * - Current post's post type is supported
	 *
	 * @since 1.2
	 * @since 1.3 Added `$post` argument.
	 *
	 * @param int|WP_Post|null $post Optional. Post ID, post object, or null to
	 *                               denote current post. Default null.
	 * @return bool
	 */
	public function can_show_ui( $post = null ) {
		return current_user_can( 'administrator' ) && in_array( get_post_type( $post ), self::get_post_types() );
	}

	/**
	 * Explicitly sets override for disabling the ability for an admin to
	 * comment on a post.
	 *
	 * @since 1.0
	 *
	 * @param bool $can_comment Should admins be able to comment on this post?
	 * @param int  $post_id     The post ID.
	 */
	public function set_admin_can_comment_on_post( $can_comment, $post_id ) {
		$comments_disabled = $this->is_admin_commenting_disabled( $post_id, false );

		// If trying to enable/disable admin commenting and the opposite state
		// is currently in effect, then make a change.
		if ( $can_comment === $comments_disabled ) {
			if ( $can_comment ) {
				delete_post_meta( $post_id, self::$setting_name );
			} else {
				update_post_meta( $post_id, self::$setting_name, '1' );
			}
		}
	}

	/**
	 * Saves the setting to explicitly disable admin commenting.
	 *
	 * @since 1.0
	 *
	 * @param int $post_id The post ID.
	 */
	public function save_setting( $post_id ) {
		if ( ! $this->can_show_ui( $post_id ) ) {
			return;
		}

		$disable_admin_commenting = isset( $_POST[ self::$setting_name ] ) && '1' === $_POST[ self::$setting_name ];
		$this->set_admin_can_comment_on_post( ! $disable_admin_commenting, $post_id );
	}

	/**
	 * Displays option field  for a post/page, but only for administrators.
	 *
	 * @since 1.0
	 *
	 * @param WP_Post|null Post or null to denote current post. Default null.
	 */
	public function display_option( $post = null ) {
		if ( ! $this->can_show_ui( $post ) ) {
			return;
		}

		if ( ! $post ) {
			$post = get_post();
		}

		$status = $this->is_admin_commenting_disabled( $post->ID, false ) ? '1' : '0';
		echo <<<HTML
			<style>
				.always-allow-admin-comments-container,
				.always-allow-admin-comments-container .description {
					display: block;
				}
				#postbox-container-1 .always-allow-admin-comments-container .description,
				.edit-post-sidebar #postbox-container-2 .always-allow-admin-comments-container .description {
					margin-top: 1em;
				}
				#postbox-container-2 .always-allow-admin-comments-container .description {
					padding-left: 1.85em;
				}
				.edit-post-sidebar #postbox-container-2 .always-allow-admin-comments-container .description {
					padding-left: 0;
				}
			</style>
HTML;
		printf( '<label for="%s" class="selectit always-allow-admin-comments-container">', esc_attr( self::$setting_name ) );
		printf( '<input type="checkbox" name="%s" value="1" %s/> ', esc_attr( self::$setting_name ), checked( $status, '1', false ) );
		_e( 'Prevent administrators from commenting.', 'always-allow-admin-comments' );
		echo '<span class="description">';
		_e( 'Administrators can otherwise comment even if commenting is closed.', 'always-allow-admin-comments' );
		echo '</span></label>' . "\n";
	}

	/**
	 * Enables admin commenting if a post's comments are closed, unless
	 * explicitly prevented from doing so.
	 *
	 * @since 1.0
	 *
	 * @param  bool $status  If the post's comments are open or closed
	 * @param  int  $post_id The post ID
	 *
	 * @return bool Whether the post's comments are open or closed
	 */
	public function comments_open_for_admin( $status, $post_id ) {
		// Only do anything if commenting is closed and the current user is an
		// Administrator.
		if ( ! $status && current_user_can( 'administrator' ) ) {
			// Enable admin commenting unless it is specifically disabled.
			if ( ! $this->is_admin_commenting_disabled( $post_id ) ) {
				$status = true;
			}
		}

		return $status;
	}

	/**
	 * Checks if a specific post has explicitly disabled admin commenting.
	 *
	 * First checks for a custom field, which gets set via a checkbox in the
	 * Discussion metabox for a post. Then runs a filter to allow programmatic
	 * override of the value (unless told not to do so by its second argument).
	 *
	 * NOTE: This doesn't first check if admin commenting is otherwise enabled
	 * or even necessary for the post. It simply checks if admin commenting has
	 * been explicitly disabled.
	 *
	 * @since 1.0
	 *
	 * @param  int  $post_id      The post ID
	 * @param  bool $apply_filter Should the filter be applied?
	 *
	 * @return bool If admin commenting has been disabled for the post.
	 */
	public function is_admin_commenting_disabled( $post_id, $apply_filter = true ) {
		$status = '1' === get_post_meta( $post_id, self::$setting_name, true );

		if ( $apply_filter ) {
			/**
			 * Filters whether admin commenting has been explicitly disabled for the
			 * given post.
			 *
			 * @since 1.0
			 *
			 * @param bool $status  Is admin commenting explicitly disabled for the post?
			 * @param int  $post_id The post ID.
			 */
			$status = apply_filters( 'c2c_always_allow_admin_comments_disable', $status, $post_id );
		}

		return $status;
	}

}

add_action( 'plugins_loaded', array( 'c2c_AlwaysAllowAdminComments', 'get_instance' ) );

endif;
